brushfire-core/src/main/scala/com/stripe/brushfire/AnnotatedTree.scala

```diff
@@ -1,25 +1,50 @@
 package com.stripe.brushfire
 
 import com.twitter.algebird._
+import com.stripe.bonsai.{ FullBinaryTree, FullBinaryTreeOps }
+
+import java.lang.Math.{abs, max}
+
+sealed abstract class Node[K, V, T, A] {
 
-sealed trait Node[K, V, T, A] {
   def annotation: A
+
+  def renumber(nextId: Int): (Int, Node[K, V, T, A]) =
+    this match {
+      case SplitNode(k, p, lc0, rc0, a) =>
+        val (n1, lc) = lc0.renumber(nextId)
+        val (n2, rc) = rc0.renumber(n1)
+        (n2, SplitNode(k, p, lc, rc, a))
+      case LeafNode(_, target, annotation) =>
+        (nextId + 1, LeafNode(nextId, target, annotation))
+    }
+
+  def fold[B](f: (Node[K, V, T, A], Node[K, V, T, A], (K, Predicate[V], A)) => B, g: Tuple3[Int, T, A] => B): B =
```

Comment:
No strong preference, more curiosity. It seems a bit weird to use a tuple if you know nothing about bonsai, but we are using it, so I think this is fine.